[BD-27] [BB-2680] Polish and prepare cc2olx converter

0x29a opened the pull request:
This PR lays ground for rework of https://github.com/openedx/cc2olx tool.

1. Added `setup.py` configuration, so package now can be installed and used like `cc2olx -i <FILE>`.
1. Added `tox` configuration, so all tests and linting can be run with `tox`.
1. Removed usage of raw `os.path` favoring `pathlib`.
1. Slightly updated readme.
1. Added `Makefile` with some utility commands.
1. Simplified `argparse` interface to have only `--inputs` argument, for both directories and files.
1. Reformatted code and removed some unused lines.
1. Removed `Cartridge.serialize` method. I'm not sure about it, but it seems that methods from `olx` module do the same.
1. Implemented zip output mode.
1. Covered mostly everything with tests.


**JIRA tickets**:
- [BLENDED-406](https://openedx.atlassian.net/browse/BLENDED-406)


**Dependencies**: None

**Testing instructions**:

1. Check that all tests and linting are passing:
    ```sh
    git clone https://github.com/edx/cc2olx
    cd cc2olx
    make test-all
    ```

2. Check that tool converts some of the test `.imscc` files without errors: 
    ```sh
    virtualenv -p python3 .fenv
    . .fenv/bin/activate
    python setup.py install
    cc2olx -i test_data/single-page.imscc
    ```

    After this, you should see new `output` directory in your current directory. It should contain `single-page.tar.gz` file.

3. Open any course in studio, then click on `Tools` -> `Import` and try to import this `single-page.tar.gz`. You should see, that your course is replaced with new imported course.


**Author notes and concerns**:

1. As @kaizoku mentioned, I'm not sure if sample `.imscc` files can be distributed.
1. I'm not sure what's the purpose of `zip` output mode, so I archive whole workspace directory.

**Reviewers**
- [ ] @kaizoku
- [ ] @staubina

edx-webhook commented:
Thanks for the pull request, @0x29a! I've created [BLENDED-406](https://openedx.atlassian.net/browse/BLENDED-406) to keep track of it in Jira.

When this pull request is ready, tag your edX technical lead.

<!-- comment:welcome-blended -->
<!-- jenkins ok to test -->

staubina commented:
Hello, Please let me know when this PR is ready for review.

kaizoku commented on src/cc2olx/main.py:
```diff
@@ -0,0 +1,47 @@
+    with open(olx_filename, "w") as olxfile:
+        olxfile.write(xml)
+
+    tgz_filename = cartridge.directory.parent / (
```
Would it make sense to put this into a differently named folder? `tmp` insinuates to me that it's disposable data, not the primary output of the program.

Maybe we could output into a folder named `output`, or into the current working directory for a single tar.gz file.

kaizoku commented on src/cc2olx/cli.py:
```diff
@@ -0,0 +1,43 @@
+        "logs. It can take one of the following values, DEBUG, INFO, WARNING, "
+        "ERROR, CRITICAL as argument.",
+    )
+    parser.add_argument(
```
Is this still in progress? I don't see the portion that utilizes this setting.

0x29a replied:
@kaizoku
>Is this still in progress?

Nope. There are no code that utilizes this setting in original repo. Should I implement it in this ticket?

kaizoku replied:
@0x29a, it'd be a bit weird to have the setting if it doesn't change anything. If we could implement this here it'd be nice.

kaizoku commented on README.rst:
```diff
@@ -0,0 +1,60 @@
+cc2olx
+######
+
+A hackathon project to explore importing Common Cartridge courses into Open edX
```
Could we update this section since we're filling out the features of this and building into more than just a hackathon project? :)

kaizoku commented on setup.py:
```diff
@@ -0,0 +1,44 @@
+        "Topic :: Utilities",
+    ],
+    description="Command line tool, that converts Common Cartridge "
+    "courses to Studio imports.",
```
Could we specify "Open edX Studio Imports" here?

kaizoku commented on setup.py:
```diff
@@ -0,0 +1,44 @@
+    setup_requires=[],
+    test_suite="tests",
+    tests_require=[],
+    url="https://github.com/openedx/cc2olx",
```
This could be the new repository URL as well.

kaizoku commented on src/cc2olx/olx.py:
```diff
@@ -0,0 +1,102 @@
+            if "children" in dd:
+                self._add_olx_nodes(child, dd["children"], tags[1:])
+
+    def _create_lti_node(self, details):
```
Coverage shows this function isn't hit. Could we add a test case or an LTI component to our test course to hit this function as well?

0x29a commented:
@kaizoku 
>One question I have is around files in test_data. Are those licensed course exports?

I found that we can distribute `allyworkshop.imscc` and `ushistory.imscc` ([look here at the bottom](https://common-cartridge-viewer.netlify.app)). But I'm not sure about the rest of them.

Probably we can find info about their licenses [here](https://lor.instructure.com/), but I can't sign up.

kaizoku commented:
Hmm, I'm not sure about the licenses for those other courses either. It may depend on the course creator.
Maybe we're better off not including them here, but adding a hyperlink to courses people can test with.

Otherwise, this works well for me @0x29a :+1:
Thanks for increasing coverage on the tests.

- [✓] I tested this: (converted a few test common cartridges, examined tests, ran tests)
- [✓] I read through the code
- ~~[n/a] I checked for accessibility issues~~
- [✓] Includes documentation

staubina commented:
I would advocate for not including test data if we do not have the license.  Maybe we can work with our content teams internally if we need to for more test data.

staubina commented on README.rst:
```diff
@@ -0,0 +1,57 @@
+-----
+
+We would love help building out this tool.  If you can help, either with
+development, or with real-world testing, please get in touch: ned@edx.org.
```
I think we might want to remove Ned's email here, but I will check in with the Open edX team.

stvstnfrd replied:
While Ned is a rock, no individual contributor should _ever_ be a single point of failure.

This was a note from when this was a hackathon project that we did _not_ expect would get official edx support.

staubina commented on src/cc2olx/models.py:
```diff
@@ -0,0 +1,625 @@
+        if not organization:
+            return
+        identifier = organization.get("identifier", "org_1")
+        # structure = organization.get("structure", "rooted-hierarchy")
```
Is this commented out code left here for a purpose?

0x29a replied:
Ah, that's from original codebase. Thanks, removed.

staubina commented on src/cc2olx/models.py:
```diff
@@ -0,0 +1,625 @@
+        if not course_root:
+            return
+        sections = course_root.get("children", [])
+        # print(
```
Same comment as before.  Do we need this commented out code?

staubina commented on src/cc2olx/models.py:
```diff
@@ -0,0 +1,625 @@
+        return output
+
+    def get_resource_content(self, identifier):
+        """
```
Could we in general add more Docstrings to the methods and files to make it easier to understand the purposes of the files and methods included? 

I can also agree to have this added in future PRs. I know this is PR is mostly a port of the existing code.

0x29a replied:
>Could we in general add more Docstrings to the methods and files to make it easier to understand the purposes of the files and methods included?

That's reasonable :+1: 
I'll definitely add more documentation.

>I can also agree to have this added in future PRs.

It would be nice to do this in following feature PR's because:
- Some code here likely will be refactored.
- I'm in progress of mastering the `imscc` and `olx` specifications, so probably can write better docs soon.

staubina commented on src/cc2olx/models.py:
```diff
@@ -0,0 +1,625 @@
+        return text
+
+    def get_title(self):
+        # TODO: Choose a better default course title
```
Just adding a note of this TODO.  Not sure if it is from current work or copied.

0x29a replied:
These TODO's are from original repo. [Source](https://github.com/openedx/cc2olx/blob/f395393e7e6f661cdeb5e28c838940de00e247ac/src/cc2olx/models.py#L310).

staubina replied:
Ok so updated in the future refactoring?

0x29a commented:
@kaizoku @staubina 
I removed `test_data` directory and added `Test Data` section in `README.rst` ([see it rendered](https://github.com/open-craft/cc2olx/tree/0x29a/bb2680/polish_cc2olx)). When fixture will contain more different types of content, we can build our own `.imscc` example from it.

Is that acceptable?

stvstnfrd commented on README.rst:
```diff
@@ -0,0 +1,68 @@
+We would love help building out this tool.  If you can help, either with
+development, or with real-world testing, please get in touch: ned@edx.org.
+
+Work that needs to be done:
```
Is this from the original `README` or is this a new note?

stvstnfrd commented on setup.py:
```diff
@@ -0,0 +1,44 @@
+        "Programming Language :: Python",
+        "Programming Language :: Python :: 3",
+        "Programming Language :: Python :: 3.5",
+        "Programming Language :: Python :: 3.6",
```
Does this actually support py3.6, py3.7?

Only 3.5 is supported at edx, while working to add support for 3.8.

stvstnfrd commented on setup.py:
```diff
@@ -0,0 +1,44 @@
+        "Topic :: Utilities",
+    ],
+    description="Command line tool, that converts Common Cartridge "
+    "courses to Open edX Studio Imports.",
```
This doesn't look like it conforms to our style guide; instead maybe:

```py
description=(
    "Command line tool that converts Common Cartridge "
    "courses to Open edX Studio imports"
)
```

Also a nit: `Imports" isn't a proper noun here; it should be lowercase.

stvstnfrd commented on src/cc2olx/cli.py:
```diff
@@ -0,0 +1,43 @@
+        "that contain them.",
+    )
+    parser.add_argument(
+        "-ll",
```
By convention, at least with GNU/linux systems (and definitely everywhere in the edx ecosystem), short args are only a single letter; can we make this `-l`?

stvstnfrd commented on src/cc2olx/models.py:
```diff
@@ -0,0 +1,619 @@
+
+
+def pprint(level, key, value, count=0):
+    return  # Comment this out to see the structure.
```
If this function is a no-op, we should remove.

stvstnfrd commented on src/cc2olx/models.py:
```diff
@@ -0,0 +1,619 @@
+            return
+        identifier = organization.get("identifier", "org_1")
+        # An organization may have a title element.
+        # title = None
```
Why is this commented?
It was not in the original.

stvstnfrd commented:
![where have all the commits gone?!](https://i.imgflip.com/48h5zp.jpg)

> I had some minor questions and comments, *but with the initial PR being so large it is hard to really pick apart.*

Yeah, I'm unclear why this PR is presented as entirely new files/commits, and not just commits on top of the existing repo.
Why have those commits disappeared, if this is _not_ a rewrite, but just a "polish and prepare"?
- Commits contain useful context; this is lost here (eg: I was able to better understand _your_ PR by analyzing individual commits)
- Authorship/accrediting is lost; now Ned, DSG, and my contributions are lost and it just looks like @0x29a wrote this from scratch.

Worse than metadata, this is expensive to review in a meaningful way. Not only is it hard for @staubina to pick this apart, but it's hard for _me_ to read and I wrote much of it originally...

Besides the debug relics that have already been pointed out, I'm still seeing a few places where debug statements and commented code remain in place. Even more confusing, is that at least one of these was added _during_ the polish and _after_ the original implementation. 

It appears that other refactors have been made, but not acknowledged in this PR; the conversion from single to double quotes is the first thing that jump outs. I can't tell what else, if anything may have changed. And no, this isn't just a rant about quotes and syntactic minutiae; it's that, "so much has changed and I can't follow it because we aren't using the tools available (commits/diffs/comments).

And that's my core feedback for this PR; it's difficult to review.
And the harder it is to review code, the more likely it is that bugs will be introduced.

So, this is kind of a long way of saying, kindly do not discard commits/history, they're valuable to reviewers and developers, now and in the future. It makes the code easier to understand, and subsequently, more solid.

And selfishly, it keeps you from needing to answer a bunch more questions from reviewers ;)

0x29a commented:
Ah, you absolutely right about commits history, @stvstnfrd. I didn't move original commits to this PR because this repo is empty, and diff would be large anyway, but yes, I completely missed authorship nuance. Sorry about this, and the fact you had to read non-optimized diff.

Luckily, that can be easily fixed. I will create another pull request with all original commits, change target of this PR to new one, and fix commits history. Does this sound good?

stvstnfrd commented:
> Ah, you absolutely right about commits history, @stvstnfrd. I didn't move original commits to this PR because this repo is empty, and diff would be large anyway, but yes, I completely missed authorship nuance. Sorry about this, and the fact you had to read non-optimized diff.
> 
> Luckily, that can be easily fixed. I will create another pull request with all original commits, change target of this PR to new one, and fix commits history. Does this sound good?

If it's not too much of a lift to rebase the history, I think it would be helpful. 

I just know from personal experience (err: failure ;) ) that these kinds of mono-commits, where new changes/refactoring squashed into original commits can hide typos and other "easy" mistakes :)

Just helps me sleep a little better :smiley: 

If the rebase gets in the weeds and is taking too much effort, I get it and feel free to stop. But hopefully the re-layering isn't too bad.

Thanks @0x29a !

0x29a commented:
@tnamgyal,

While I can rebase [original master branch](https://github.com/openedx/cc2olx/tree/master) with [this repo's master](https://github.com/edx/cc2olx/tree/master) and create pull request with original commits, IMO clean way would be hard reset to https://github.com/openedx/cc2olx/commit/f395393e7e6f661cdeb5e28c838940de00e247ac because:
- [LICENSE](https://github.com/edx/cc2olx/blob/master/LICENSE) in this repo differs from [LICENSE](https://github.com/openedx/cc2olx/blob/d480b7442f98c594703904eabb4d53567ca29980/LICENSE) in original repo.
- I want to avoid dual authorship for commits in history.

Can you please reset current master to https://github.com/openedx/cc2olx/commit/f395393e7e6f661cdeb5e28c838940de00e247ac and push with force?

Thanks.

stvstnfrd commented:
@0x29a That sounds like a good approach. I hadn't thought about the no-force-push-master issue.

I've also created an IT ticket on our end to request the reset, to help move things along.

0x29a commented:
>I've also created an IT ticket on our end to request the reset, to help move things along.

Thanks!

staubina commented:
@stvstnfrd I have access to this repos permissions. I just need to know what to do.  If you have a moment reach out today and we can get it done.

staubina commented:
We should be all set to add the history to Master now. Let me know.

staubina commented:
We updated the history here @0x29a so you will need to apply your changes on top of the History and create a new PR. Sorry for the hassle.

staubina commented:
This PR was closed forcefully by GitHub.

0x29a commented:
@staubina
>This PR was closed forcefully by GitHub.

Ah, that's probably because this branch has entirely different history.

Thanks, I'll submit new one soon If I can't reopen this one after fixing history.

staubina commented:
It is. when I try to reopen it says I cannot do it because of a history mis match.
